File: 13. Implement strStr().cpp
class Solution {
public:
    /*
     * @param source: source string to be scanned.
     * @param target: target string containing the sequence of characters to match
     * @return: a index to the first occurrence of target in source, or -1  if target is not part of source.
     */
    int strStr(const char *source, const char *target) {
        if (source == NULL || target == NULL) return -1;
        int lens = strlen(source);
        int lent = strlen(target);
        if (lens == 0 && lent == 0) return 0;
        else if (lens == 0) return -1;
        else if (lent == 0) return 0;
        for (int k = 0; k < (lens - lent) + 1; ++k) {
            if (source[k] == target[0]) {
                int temp_k = k;
                int temp_t = 0;
                while (source[temp_k] == target[temp_t] && temp_t < lent) {
                    temp_k += 1;
                    temp_t += 1;
                }
                if (temp_t == lent) return k;
            }
        }
        return -1;
    }
};

// 总耗时 199 ms
// 您的提交打败了 68.00% 的提交!

File: Python3/13. Implement strStr().py
#!/usr/bin/env python3

class Solution:
    """
    @param: source: source string to be scanned.
    @param: target: target string containing the sequence of characters to match
    @return: a index to the first occurrence of target in source, or -1  if target is not part of source.
    """
    def strStr(self, source, target):
        if source == None or target == None : return -1
        if target == '' : return 0
        if source == '' : return -1
        source_len = len(source)
        target_len  = len(target)
        print(source[0 : 0 + target_len])
        for i in range(0, source_len - target_len + 1) :
            if target == source[i : i + target_len] :
                return i
        return -1

# 总耗时 821 ms
# 您的提交打败了 99.80% 的提交!